web_pwa_cache/static/src/js/worker/components/importer.js

```diff
@@ -0,0 +1,209 @@
+
+    /**
+     * This class is used to store Odoo replies
+     * The name of the functions match with the name of the python implementation.
+     */
+    const SWImporterComponent = SWComponent.extend({
+        /**
+         * @param {String} model
+         * @param {Object} data
+         * @returns {Promise}
+         */
+        name_search: function(model, data) {
+            const records = _.map(data, record => {
+                return {id: record[0], display_name: record[1]};
+            });
+            const tasks = [];
+            for (const record of records) {
+                tasks.push(this._db.write(model, [record.id], record));
+            }
+            return Promise.all(tasks);
+        },
+
+        /**
+         * @param {String} model
+         * @param {Object} data
+         * @returns {Promise}
+         */
+        name_get: function(model, data) {
+            return this.name_search(model, data);
+        },
+
+        /**
+         * @param {String} model
+         * @param {Object} data
+         * @returns {Promise}
+         */
+        load_views: function(model, data) {
+            return new Promise(async (resolve, reject) => {
+                const fields_views = data.fields_views;
+                try {
+                    await this._db.indexeddb.views.bulkPut(fields_views);
+                } catch (err) {
+                    return reject(err);
+                }
+                return resolve();
+            });
+        },
+
+        /**
+         * @param {String} model
+         * @param {Object} data
+         * @returns {Promise}
+         */
+        default_get: function(model, data) {
+            return new Promise(async (resolve, reject) => {
+                try {
+                    const model_info_metadata = await this._db.getModelInfo(
+                        "model_metadata",
+                        true
+                    );
+                    const rc_ids = await this._db.search(model_info_metadata, [
+                        ["model", "=", model],
+                    ]);
+                    await this._db.sqlitedb.updateModelInfo(rc_ids, {
+                        defaults: data,
+                    });
+                } catch (err) {
+                    return reject(err);
+                }
+
+                return resolve();
+            });
+        },
+
+        /**
+         * @param {String} model
+         * @param {Object} data
+         * @returns {Promise}
+         */
+        load_menus: function(model, data) {
+            const values = {
+                param: "menus",
+                value: data,
+            };
+            return this._db.indexeddb.userdata.put(values);
+        },
+
+        /**
+         * @param {String} model
+         * @param {Object} data
+         * @returns {Promise}
+         */
+        read: function(model, data) {
+            return this._db.write(model, [data.id], data);
```

Comment:
Managed to end up in a state whereby the data returned on `read` was an array (just on browsing `res.partner` records in form view on a stock 13.0 demo database, nothing fancy), then came up with:

```
           if (_.isArray(data)) {
               var self = this;
               _.each(data, function(item) {
                   self._db.write(model, item.id, item);
               });
           } else {
               return this._db.write(model, [data.id], data);
           }
```

Not 100% sure yet if this is good or whether my problem originated elsewhere.